File: py/0021_MergeTwoSortedLists.py
# 21. Merge Two Sorted Lists

from typing import Optional
from classes import ListNode
from utils import toLinkedList, printLinkedList


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1 or not list2:
            return list1 if list1 else list2

        head = curr = ListNode(-10000)
        while list1 and list2:
            if list1.val < list2.val:
                curr.next = list1
                curr = curr.next
                list1 = list1.next
            else:
                curr.next = list2
                curr = curr.next
                list2 = list2.next

        if list1:
            curr.next = list1
        if list2:
            curr.next = list2

        return head.next


if __name__ == "__main__":
    puzzles = [
        ([1, 2, 4], [1, 3, 4]),
        ([], []),
        ([], [0]),
        ([1, 3, 5, 7], [4, 6, 8, 10, 12])
    ]
    for a, b in puzzles:
        printLinkedList(Solution().mergeTwoLists(
            toLinkedList(a), toLinkedList(b)))

"""
Runtime
32
ms
Beats
90.15%
Memory
16.55
MB
Beats
27.14%
"""

File: go/0021/0021_MergeTwoSortedLists.go
// 21. Merge Two Sorted Lists

package main

import (
	"bufio"
	"encoding/json"
	"leetcode/utils"
	"os"

	"github.com/samber/lo"
)

func mergeTwoLists(list1 *utils.ListNode, list2 *utils.ListNode) *utils.ListNode {
	if list1 == nil && list2 == nil {
		return nil
	} else if list1 != nil && list2 == nil {
		return list1
	} else if list1 == nil && list2 != nil {
		return list2
	}

	var result *utils.ListNode = nil
	var curr *utils.ListNode = nil
	if list1.Val < list2.Val {
		result = list1
		curr = list1
		list1 = list1.Next
	} else {
		result = list2
		curr = list2
		list2 = list2.Next
	}

	// compare and take the smallest element
	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			curr.Next = list1
			list1 = list1.Next
		} else {
			curr.Next = list2
			list2 = list2.Next
		}
		curr = curr.Next
	}

	// attach leftover elements to result
	if list1 != nil {
		curr.Next = list1
	}

	if list2 != nil {
		curr.Next = list2
	}

	return result
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	const testSize int = 2

	for _, test := range lo.Chunk(lines, testSize) {
		var first []int
		var second []int
		if err1 := json.Unmarshal([]byte(test[0]), &first); err1 != nil {
			panic(err1)
		}
		if err2 := json.Unmarshal([]byte(test[1]), &second); err2 != nil {
			panic(err2)
		}
		result := mergeTwoLists(utils.ToLinkedList(first), utils.ToLinkedList(second))
		utils.PrintLinkedList(result)
	}
}
